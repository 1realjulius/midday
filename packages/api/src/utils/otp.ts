export const generateOtp = (limit = 6) => {
  const digits = "0123456789";
  let OTP = "";

  for (let i = 0; i < limit; i++) {
    OTP += digits[Math.floor(Math.random() * 10)];
  }

  return OTP;
};

export interface SmsOptions {
  to: string;
  text: string;
}

export const sendSMS = async ({ to, text }: SmsOptions) => {
  try {
    const response = await fetch("https://rest.nexmo.com/sms/json", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        from: "Lasagne",
        to,
        text,
        api_key: process.env.VONAGE_API_KEY!,
        api_secret: process.env.VONAGE_API_SECRET!,
      }),
    });

    return response.json();
  } catch (error) {
    console.log(error);
  }
};
