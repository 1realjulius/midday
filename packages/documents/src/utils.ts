export const findValue = (entities, type: string) => {
  const found = entities.find((entry) => entry.type === type);
  return found?.normalizedValue?.text || found?.mentionText;
};
