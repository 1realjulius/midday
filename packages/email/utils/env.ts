export function getEnvironmentUrl() {
  if (process.env.NODE_ENV === "development") {
    return "http://localhost:3000/email";
  }

  return "https://midday.ai/email";
}
