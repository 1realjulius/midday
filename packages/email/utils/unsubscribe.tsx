import { SignJWT, jwtVerify } from "jose";

const secretKey = new TextEncoder().encode(process.env.UNSUBSCRIBE_JWT_SECRET!);

type GenerateUnsubscribeLinkParams = {
  id: string;
  type: "onboarding";
};

export async function generateUnsubscribeLink({
  id,
  type,
}: GenerateUnsubscribeLinkParams) {
  const token = await new SignJWT({
    id,
    type,
  })
    .setProtectedHeader({ alg: "HS256" })
    .setIssuedAt()
    .setExpirationTime("1 day")
    .sign(secretKey);

  const url = new URL(`/unsubscribe/${token}`, "https://app.midday.ai");

  url.searchParams.append("type", type);

  return url.toString();
}

export async function verifyUnsubscribeLink(token: string) {
  try {
    const { payload } = await jwtVerify(token, secretKey, {
      issuer: process.env.UNSUBSCRIBE_JWT_ISSUER,
    });

    return payload;
  } catch (e) {
    console.log("Token is invalid");
  }
}
