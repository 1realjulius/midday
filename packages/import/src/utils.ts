import crypto from "node:crypto";
import type { Transaction } from "@midday/providers/src/types";
import { capitalCase } from "change-case";
import { parse } from "csv-parse/sync";
import { formatISO } from "date-fns";
import { z } from "zod";
import type { ExtractedTransaction, FindIndexesByKeyParams } from "./types";

function generateId(value: string) {
  const hash = crypto.createHash("sha256");
  hash.update(value);

  return hash.digest("hex");
}

type TransformTransaction = Transaction & { teamId: string };

export function transformAmount(amount: string | number) {
  if (typeof amount === "string") {
    return +amount.replace(/\s/g, "").replace(",", ".");
  }

  return amount;
}

export function transform(transaction: TransformTransaction) {
  const internalId = generateId(
    `${transaction.date}-${transaction.name}-${transaction.amount}`
  );
  return {
    internal_id: internalId,
    team_id: transaction.teamId,
    status: "posted",
    date: formatISO(transaction.date, { representation: "date" }),
    amount: transformAmount(transaction.amount),
    name: transaction?.description && capitalCase(transaction.description),
    manual: true,
  };
}

export function parseCsv(input: string) {
  return parse(input, {
    delimiter: [",", ";", "."],
    skip_empty_lines: true,
    skip_records_with_empty_values: true,
    skip_records_with_error: true,
    trim: true,
    relax_column_count: true,
  });
}

export const transactionSchema = z.object({
  amount: z.string(),
  description: z.string(),
  date: z.coerce.date(),
});

export const filterTransactions = (row: ExtractedTransaction) => {
  const parsed = transactionSchema.safeParse(row);

  if (parsed.success) {
    return true;
  }

  return false;
};

export const findIndexesByKey = ({
  raw,
  input,
  key,
  parse = (value) => value.toString().trim(),
}: FindIndexesByKeyParams) => {
  const values = input.map((v) => parse(v[key]));

  const matchIndices = raw.map((r) =>
    r.findIndex((v) => values.includes(parse(v)))
  );

  return Math.max(...matchIndices);
};
