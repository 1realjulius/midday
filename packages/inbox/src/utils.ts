export function getInboxIdFromEmail(email: string) {
  return email.split("@").at(0);
}

export const findDocumentValue = (entities, type) => {
  const found = entities.find((entry) => entry.type === type);
  return found?.normalizedValue?.text || found?.mentionText;
};
