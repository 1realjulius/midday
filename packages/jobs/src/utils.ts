import { capitalCase } from "change-case";

export const mapTransactionMethod = (method: string) => {
  switch (method) {
    case "Payment":
    case "Bankgiro payment":
    case "Incoming foreign payment":
      return "payment";
    case "Card purchase":
    case "Card foreign purchase":
      return "card_purchase";
    case "Card ATM":
      return "card_atm";
    case "Transfer":
      return "transfer";
    default:
      return "other";
  }
};

export async function processPromisesBatch(
  items: Array<any>,
  limit: number,
  fn: (item: any) => Promise<any>
): Promise<any> {
  let results = [];
  for (let start = 0; start < items.length; start += limit) {
    const end = start + limit > items.length ? items.length : start + limit;

    const slicedResults = await Promise.all(items.slice(start, end).map(fn));

    results = [...results, ...slicedResults];
  }

  return results;
}

export const transformTransactions = (transactions, { teamId, accountId }) =>
  // We want to insert transactions in reversed order so the incremental id in supabase is correct
  transactions?.reverse().map((data) => ({
    transaction_id: data.transactionId,
    reference: data.entryReference,
    booking_date: data.bookingDate,
    date: data.valueDate,
    name: capitalCase(data.additionalInformation),
    original: data.additionalInformation,
    method: mapTransactionMethod(data.proprietaryBankTransactionCode),
    internal_id: data.internalTransactionId,
    amount: data.transactionAmount.amount,
    currency: data.transactionAmount.currency,
    bank_account_id: accountId,
    category: data.transactionAmount.amount > 0 ? "income" : null,
    team_id: teamId,
  }));
