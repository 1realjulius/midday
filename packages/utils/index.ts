export function stripSpecialCharacters(inputString: string) {
  // Use a regular expression to replace all non-alphanumeric characters except hyphen, space, dot,and parentheses with an empty string
  return inputString.replace(/[^a-zA-Z0-9\s.()-]/g, "");
}

export function shuffle(array: any) {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]];
  }
  return array;
}
