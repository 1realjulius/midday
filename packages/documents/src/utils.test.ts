import { expect, test } from "bun:test";
import { getDomainFromEmail } from "./utils";

test("Get domain from email", () => {
  expect(getDomainFromEmail("invoice@supabase.com")).toMatch("supabase.com");
});
