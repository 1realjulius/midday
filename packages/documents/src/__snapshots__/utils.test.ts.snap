// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate comma separated string 1`] = `
"Supplier name: Savoj, Receipt date: 2024-04-19, Purchase time: 19:25:46, Supplier address: Roslagsgatan 43
113 54 Stockholm
Sverige, Supplier city: Stockholm, Supplier phone: null, Supplier email: null, Items: 1 marinara 119,00, 1 165,00, 1 filetti 169,00, 1 pepperoni, Margherita 298,00, Nduja, Pesto Burrata, Net amount: 1149.99, Total amount: 189, Currency: SEK"
`;
